hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -3100,8 +3100,12 @@ private void registerEntityNameUsage(
 				(s, existingUse) -> finalEntityNameUse.stronger( existingUse )
 		);
 
-		// Resolve the table reference for all types which we register an entity name use for
-		if ( actualTableGroup.isInitialized() ) {
+		// Resolve the table reference for all types which we register an entity name use for.
+		// Also, force table group initialization for treats when needed to ensure correct cardinality
+		final EntityNameUse.UseKind useKind = finalEntityNameUse.getKind();
+		if ( actualTableGroup.isInitialized() || ( useKind == EntityNameUse.UseKind.TREAT
```

Comment:
I guess `canUseInnerJoin()` might be a good indicator to understand if the treat is necessary to retain cardinality. Did you try that before?

```suggestion
		if ( actualTableGroup.isInitialized() || ( useKind == EntityNameUse.UseKind.TREAT && actualTableGroup.canUseInnerJoins()
```

Reply:
Yes checking `canUseInnerJoin()` makes sense, it should help avoid unnecessary initializations and type casts, I've added that to the condition - thanks.

By the way, the `instanceof EntityMappingType` check might be unnecessary, I added it to be safe.

Reply:
Indeed, the `instanceof` check is unnecessary as this is only ever called for table groups for entity types.

Reply:
Gone :+1: